### PR TITLE
runtime(netrw): fixed several bugs in tree listing

### LINKED ISSUE #5630: netrw confuses directories when netrw_liststyle is 3
When you have multiple directories expanded in the tree view (netrw_liststyle set to 3), netrw will try to open whatever filename you select from the last selected directory instead of the directory it is currently in. For example suppose you have the following files:

foo/file1.txt
foo/bar/file2.txt

Navigate to open foo/bar/file2.txt. Then go back to the tree view and attempt to open foo/file1.txt. It will try to open foo/bar/file1.txt instead of foo/file1.txt. The workaround is to either try to reopen it or select the correct directory in advance and only then open your file.

This bug seems super dangerous.

### LINKED ISSUE #9807: netrw folders vanish and lose trailing slash in tree-view after focus gained
### Steps to reproduce

0. build latest master with default configure options

1. set minimal vimrc 
```
set nocompatible
let g:netrw_liststyle=3
```

2. create test folders
```
mkdir -p a/b/c
mkdir -p a/b/c2
mkdir -p a/b/c3/d
```

 2.5. have test folder structure
```
  a
  └── b
      ├── c
      ├── c2
      └── c3
          └── d
```

3. open netrw in folder a:
```
:Lexplore a
```

4. enter folder b (with `<cr>`)

5.  enter folder c3 and have this tree (header omitted):
```
../
a/
| b/
| | c/
| | c2/
| | c3/
| | | d/
```

6.  De-select the terminal (lose focus)
7.  Select vim again (regain focus)

8. See this (header omitted again):
```
../
a/
| b/
| | c
| | c2
| | c3
```
(Clicking those folders without slashes opens another netrw view with that folder as root - but that part is, in that broken state, technically correct I would say)

Note: it only happens when you are "3 folders deep" - without opening 'd' it does not break.
Note2: Closing and re-opening folder 'b' "fixes" it - until the next focus lost/regain cycle.

### Expected behaviour

Folders should not lose their trailing slash or vanish :)

### Version of Vim

8.2.4424

### Environment

Tested on Darwin arm (macos 12.2.1) and Ubuntu 21.04
xterm-256color
zsh 5.8
Tested in iTerm2 Build 3.5.git.ebe2bf0883 and Alacritty 0.1.0.0-rc4 (of course, both with focus reporting enabled) 

### Logs and stack traces

_No response_

### LINKED ISSUE #14623: Wrong handling of folder symbolic link in tree view
### Steps to reproduce

1. Download latest vim appimage.
2. `mv Vim-v9.1.0366.glibc2.29-x86_64.AppImage vim`
3. `mkdir /tmp/test`
4. `mv vim /tmp/test && chmod u+x /tmp/test/vim`
5. `cd /tmp/test && mkdir first && echo 'text' >> first/file.txt`
6. `ln -sv /tmp/test/first /tmp/test/second`
7. `./vim -c 'let g:netrw_liststyle=3'`
8. Try to open `second` folder which is a symlink. For some reason `netrw` added extra `second` to it's path, and turned it into `/tmp/test/second/second`, when it should have been `/tmp/test/second`

https://github.com/vim/vim/assets/25168308/527c75df-110c-4894-a12f-153edef872f1

### Expected behaviour

Open a symbolic link

### Version of Vim

VIM - Vi IMproved 8.2 (2019 Dec 12, compiled Mar 14 2024 09:05:11)

### Environment

Ubuntu 22.04.4 LTS jammy
GNOME Terminal 3.44.0 using VTE 0.68.0 +BIDI +GNUTLS +ICU +SYSTEMD
xterm-256color
zsh 5.8.1 (x86_64-ubuntu-linux-gnu)

### Logs and stack traces

_No response_

### PR DESCRIPTION
**Pull request by tomben13**
Bugs fixed when using tree listing style :

- Hitting Ctrl-L causes directories to appear as files, fixes #9807 
- Cursor would jump to the top when opening a directory. Fixed one case of this issue, it can still happen under certain circumstances
- When opening a symlink, netrw would use b:netrw_curdir as the base directory for the symlink but the symlink might not be under b:netrw_curdir because of the tree view. Should fix #14623. Might also fix #5630 
- Symlink resolve didn't work after refresh sometimes in `s:ShowLink`.

Added an argument to `s:NetrwBrowseChgDir` : `cursor` forces the `dirname` value to be retrieved from the word under the cursor. If `cursor` is 1 and tree view, `dirname` is computed with `s:NetrwTreePath`. Wherever there is `s:NetrwBrowseChgDir(.., s:NetrwGetWord(),..)`, `cursor` was set to 1

Known remaining issue :
- Hitting Ctrl-L no longer causes directories to appear as files but sorting is incorrect for the files that are one level deep in the tree listing

### CODE REVIEW

File: runtime/autoload/netrw.vim [hunk 1]
```diff
@@ -4758,7 +4765,7 @@ endfun
 "                       directory and a new directory name.  Also, if the
 "                       "new directory name" is actually a file,
 "                       NetrwBrowseChgDir() edits the file.
-fun! s:NetrwBrowseChgDir(islocal,newdir,...)
+fun! s:NetrwBrowseChgDir(islocal,newdir,cursor,...)
```
Comment (chrisbra) [inline runtime/autoload/netrw.vim:4771]: can you please add a comment what the purpose of the `cursor` arg is?

File: runtime/autoload/netrw.vim [hunk 2]
```diff
@@ -9518,7 +9534,7 @@ fun! s:NetrwTreeDisplay(dir,depth)
     NetrwKeepj call s:NetrwTreeDisplay(direntry.'/',depth)
    elseif entry =~ '@$' && has_key(w:netrw_treedict,direntry.'@')
 "    call Decho("<".direntry."/> is a key in treedict - display subtree for it",'~'.expand("<slnum>"))
-    NetrwKeepj call s:NetrwTreeDisplay(direntry.'/',depth)
+    NetrwKeepj call s:NetrwTreeDisplay(direntry.'@',depth)
```
Comment (chrisbra) [inline runtime/autoload/netrw.vim:9540]: why the change of `/` to `@`?

File: runtime/autoload/netrw.vim [hunk 3]
```diff
@@ -12523,14 +12543,19 @@ fun! s:ShowLink()
 " "  call Decho(printf("line#%4d: %s",line("."),getline(".")),'~'.expand("<slnum>"))
   if exists("b:netrw_curdir")
    norm! $?\a
-   let fname   = b:netrw_curdir.'/'.s:NetrwGetWord()
+   if exists("w:netrw_liststyle") && w:netrw_liststyle == s:TREELIST && exists("w:netrw_treetop")
+    let basedir = s:NetrwTreePath(w:netrw_treetop)
+   else
+    let basedir = b:netrw_curdir.'/'
+   endif
+   let fname = basedir .. s:NetrwGetWord()
```
Comment (chrisbra) [inline runtime/autoload/netrw.vim]:
```suggestion
   let fname = basedir . s:NetrwGetWord()
```

File: runtime/autoload/netrw.vim (hunk 2 above)
Comment (tomben13) [reply runtime/autoload/netrw.vim:9540]: When `has_key(w:netrw_treedict,direntry.'@')`, we must call `s:NetrwTreeDisplay` with the correct name of the key which is `direntry.'@'`

File: runtime/autoload/netrw.vim (hunk 3 above)
Comment (tomben13) [reply runtime/autoload/netrw.vim]: From `:h expr-.` : 
_For String concatenation ".." is preferred, since "." is ambiguous, it is also
used for |Dict| member access and floating point numbers._

I can still use `.` to be consistent with the rest of the code base if you prefer.

File: runtime/autoload/netrw.vim (hunk 1 above)
Comment (tomben13) [reply runtime/autoload/netrw.vim:4771]: Done. Is it clear enough ?

File: runtime/autoload/netrw.vim (hunk 2 above)
Comment (chrisbra) [reply runtime/autoload/netrw.vim:9540]: so the use of `/` was an error?

File: runtime/autoload/netrw.vim (hunk 3 above)
Comment (chrisbra) [reply runtime/autoload/netrw.vim]: Yeah, I am mainly worried about consistency here. the lines around tend to use `.` for concatenation, so I'd prefer if we keep the style.
(or we can change the style per functions, but I would prefer to have this as a followup clean up PR then)

File: runtime/autoload/netrw.vim (hunk 1 above)
Comment (chrisbra) [reply runtime/autoload/netrw.vim:4771]: yes thanks

File: runtime/autoload/netrw.vim (hunk 2 above)
Comment (tomben13) [reply runtime/autoload/netrw.vim:9540]: I think yes. But actually this case probably never happens. As links are usually not listed with trailing `@` in `w:netrw_treedict`. But maybe there are cases where it can happen that I'm not aware of. I think in this case this was an error.

**Comment by chrisbra** [comment]
Thanks I include it and clean up a few of those `Decho` commands as well while at it
